File: src/main/java/kr/co/bookand/backend/notice/service/NoticeService.java
package kr.co.bookand.backend.notice.service;

import kr.co.bookand.backend.account.service.AccountService;
import kr.co.bookand.backend.common.domain.Status;
import kr.co.bookand.backend.common.domain.dto.PageResponse;
import kr.co.bookand.backend.common.exception.ErrorCode;
import kr.co.bookand.backend.notice.domain.Notice;
import kr.co.bookand.backend.notice.domain.dto.NoticeDto.*;
import kr.co.bookand.backend.notice.exception.NoticeException;
import kr.co.bookand.backend.notice.repository.NoticeRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
@Slf4j
public class NoticeService {

    private final NoticeRepository noticeRepository;
    private final AccountService accountService;

    @Transactional
    public NoticeIdResponse createNotice(CreateNoticeRequest createNoticeRequest) {
        accountService.isAccountAdmin();
        Notice notice = createNoticeRequest.toEntity();
        noticeRepository.save(notice);
        return new NoticeIdResponse(notice.getId());
    }

    public PageResponse<NoticeResponse> getNoticeSimpleList(Pageable pageable, Long cursorId) {
        Page<NoticeResponse> notificationList = noticeRepository.findAllByVisibilityAndStatus(pageable, true, Status.VISIBLE, cursorId)
                .map(NoticeResponse::of);
        return PageResponse.of(notificationList);
    }

    public NoticeResponse getNotice(Long notificationId) {
        Notice notice = noticeRepository.findById(notificationId)
                .orElseThrow(() -> new NoticeException(ErrorCode.NOT_FOUND_NOTIFICATION, notificationId));
        return NoticeResponse.of(notice);
    }

    public PageResponse<NoticeWebResponse> getNoticeList(Pageable pageable) {
        accountService.isAccountAdmin();
        return PageResponse.of(noticeRepository.findAll(pageable).map(NoticeWebResponse::of));
    }

    @Transactional
    public NoticeIdResponse updateNotice(Long notificationId, UpdateNoticeRequest updateNoticeRequest) {
        accountService.isAccountAdmin();
        Notice notice = findNotice(notificationId);
        notice.updateNoticeData(updateNoticeRequest);
        return new NoticeIdResponse(notice.getId());
    }

    @Transactional
    public NoticeMessage deleteNotice(Long notificationId) {
        accountService.isAccountAdmin();
        Notice notice = findNotice(notificationId);
        notice.softDelete();
        notice.updateNoticeStatus(Status.REMOVE);
        return new NoticeMessage("Delete Success.");
    }

    @Transactional
    public NoticeMessage updateNoticeStatus(Long notificationId) {
        accountService.isAccountAdmin();
        Notice notice = findNotice(notificationId);
        String message;
        if (notice.getStatus().equals(Status.VISIBLE)) {
            notice.updateNoticeStatus(Status.INVISIBLE);
            message = "Update Status to Invisible.";
        } else {
            notice.updateNoticeStatus(Status.VISIBLE);
            message = "Update Status to Visible.";
        }
        notice.updateDisplayAt(LocalDateTime.now());
       return new NoticeMessage(message);
    }

    public Notice findNotice(Long notificationId) {
        Notice notice = noticeRepository.findById(notificationId)
                .orElseThrow(() -> new NoticeException(ErrorCode.NOT_FOUND_NOTIFICATION, notificationId));
        return notice;
    }
}

File: src/main/kotlin/kr/co/bookand/backend/notice/service/NoticeService.kt
package kr.co.bookand.backend.notice.service

import kr.co.bookand.backend.account.model.Account
import kr.co.bookand.backend.common.ErrorCode
import kr.co.bookand.backend.common.PageResponse
import kr.co.bookand.backend.common.Status
import kr.co.bookand.backend.common.exception.BookandException
import kr.co.bookand.backend.common.model.MessageResponse
import kr.co.bookand.backend.notice.dto.*
import kr.co.bookand.backend.notice.model.Notice
import kr.co.bookand.backend.notice.repository.NoticeRepository
import lombok.RequiredArgsConstructor
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
class NoticeService(
    private val noticeRepository: NoticeRepository
) {

    @Transactional
    fun createNotice(
        currentAccount: Account,
        createNoticeRequest: CreateNoticeRequest
    ): NoticeIdResponse {
        currentAccount.role.checkAdminAndManager()
        val notice = Notice(createNoticeRequest)
        val saveNotice = noticeRepository.save(notice)
        return NoticeIdResponse(saveNotice.id)
    }

    fun getNoticeSimpleList(
        pageable: Pageable,
        cursorId: Long?
    ): PageResponse<NoticeResponse> {
        val nextCursorId = cursorId ?: 0L
        val page = noticeRepository.findAllByVisibilityAndStatus(pageable, true, Status.VISIBLE, nextCursorId)
            .map { NoticeResponse(it) }
        return PageResponse.of(page)
    }

    @Transactional
    fun updateNotice(
        currentAccount: Account,
        noticeId: Long,
        updateNoticeRequest: UpdateNoticeRequest
    ): NoticeIdResponse {
        currentAccount.role.checkAdminAndManager()
        val notice = getNotice(noticeId)
        notice.updateNoticeData(updateNoticeRequest)
        return NoticeIdResponse(notice.id)
    }

    @Transactional
    fun deleteNotice(currentAccount: Account, noticeId: Long): MessageResponse {
        currentAccount.role.checkAdminAndManager()
        noticeRepository.deleteById(noticeId)
        return MessageResponse(result = "공지 삭제 완료", statusCode = 200)
    }

    @Transactional
    fun updateNoticeStatus(
        currentAccount: Account,
        noticeId: Long
    ): NoticeIdResponse {
        currentAccount.role.checkAdminAndManager()
        val notice = getNotice(noticeId)
        val status = notice.status
        if (status == Status.VISIBLE) notice.updateNoticeStatus(Status.INVISIBLE)
        else {
            notice.updateNoticeStatus(Status.VISIBLE)
            notice.updateDisplayAt()
        }
        return NoticeIdResponse(notice.id)
    }

    fun getNoticeDetail(noticeId: Long): NoticeResponse {
        val notice = getNotice(noticeId)
        return NoticeResponse(notice)
    }

    fun getNoticeList(
        pageable: Pageable
    ): NoticePageResponse {
        val list = noticeRepository.findAllByVisibility(pageable, true)
            .map { NoticeResponse(it) }
        return NoticePageResponse(PageResponse.of(list))
    }

    fun getNotice(noticeId: Long): Notice {
        return noticeRepository.findById(noticeId)
            .orElseThrow { throw BookandException(ErrorCode.NOT_FOUND_NOTIFICATION) }
    }


}